src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -472,43 +481,176 @@ public void onClusterServiceClose() {
                 @Override
                 public void onTimeout(TimeValue timeout) {
                     // Try one more time...
-                    doStart();
+                    start();
                 }
             });
         }
 
-        void performOnPrimary(int primaryShardId, final ShardRouting shard) {
+        /** upon success, finish the first phase and transfer responsibility to the {@link ReplicationPhase} */
+        void finishAndMoveToReplication(ReplicationPhase replicationPhase) {
+            if (finished.compareAndSet(false, true)) {
+                replicationPhase.start();
+            } else {
+                assert false : "finishAndMoveToReplication called but operation is already finished";
+            }
+        }
+
+
+        void finishAsFailed(Throwable failure) {
```

Comment:
my main concern was the lack of separation between the part that sends remote replication request and the rest of the code. This is what I separated here. In the future we will remove all local execution and use the short cut in the transport service